### PR TITLE
fix: token balance load

### LINKED ISSUE #2386: Android: Home Screen balance loading bug
On android, we display first a 0, then a -, and the balance when the app is cold started. On iOS, Nitya says she doesn't see this.

### PR DESCRIPTION
**Pull request by MuckT**
### Description

A fix for when stale prices are present in the redux store causing user balances to display $0.00 when balances are stale.

### Other changes

Converts the token reducer to a slice.

### Tested

Tested locally on iOS and Android.

### How others should test

1. With an account with a balance Open the App the app and load the prices.
2. Prices should load.
3. Background the app and wait for the cache to expire, 24 hours by default, but can be edited in by adjusting `TIME_UNTIL_TOKEN_INFO_BECOMES_STALE` in the `config.ts` to emulate this behavior for reviews.
4. Foreground the App. 
5. Total Balance should display as '-' while loading or if the prices are stale. Accounts with balances should never display $0.00 balance.

### Related issues

- Fixes #2386

### Backwards compatibility

Yes

### CODE REVIEW

**Comment by MuckT** [comment]
@kathaypacific it looks like this failing e2e tests in CI around 50% of the time: I'll take another look.

File: src/components/TokenBalance.test.tsx [hunk 1]
```diff
@@ -120,7 +121,17 @@ describe('FiatExchangeTokenBalance and HomeTokenBalance', () => {
     async (TokenBalanceComponent) => {
       const store = createMockStore({
         tokens: {
-          tokenBalances: {},
+          tokenBalances: {
```
Comment (kathaypacific) [inline src/components/TokenBalance.test.tsx:124]: just checking, without this change does this test still pass? for a new install, the initial value for `tokenBalances` is `{}` so would be nice to test this base case

File: src/redux/store.test.ts [hunk 2]
```diff
@@ -329,7 +329,7 @@ describe('store state', () => {
         },
         "tokens": Object {
           "error": false,
-          "loading": false,
+          "loading": true,
```
Comment (kathaypacific) [inline src/redux/store.test.ts:332]: is it possible to not update this value? usually the initial state for loading should be false?

File: src/tokens/selectors.ts [hunk 3]
```diff
@@ -50,6 +50,22 @@ export const tokensWithUsdValueSelector = createSelector(tokensListSelector, (to
...
+  if (tokensWithUsdValue.length !== 0) {
+    return (
+      tokensWithUsdValue.filter(
+        (tokenInfo) =>
+          (tokenInfo.priceFetchedAt ?? 0) < Date.now() - TIME_UNTIL_TOKEN_INFO_BECOMES_STALE
+      ).length > 0
+    )
+  } else {
+    return true
+  }
```
Comment (kathaypacific) [inline src/tokens/selectors.ts:66]: it took my brain a little time to understand this, do you think this alternative is the same/easier to read?
```
return tokens.every(tokenInfo => tokenInfo.usdPrice === null) || tokens.some(tokenInfo => (tokenInfo.priceFetchedAt ?? 0) < Date.now() - TIME_UNTIL_TOKEN_INFO_BECOMES_STALE)
```

**Review by kathaypacific — COMMENTED** [review]
Hey this is looking great!!!!

File: src/redux/store.test.ts (hunk 2 above)
Comment (MuckT) [reply src/redux/store.test.ts]: Fixed in https://github.com/valora-inc/wallet/pull/2645/commits/8aa0b731297bb537e31822c32d61d08447f38aa0!

File: src/components/TokenBalance.test.tsx (hunk 1 above)
Comment (MuckT) [reply src/components/TokenBalance.test.tsx]: Fixed in https://github.com/valora-inc/wallet/pull/2645/commits/653519e04f1d418a0b98c79679747670e0e89868!

File: src/tokens/selectors.ts (hunk 3 above)
Comment (MuckT) [reply src/tokens/selectors.ts:68]: Tried the exact code and prices where showing as stale when not expected. Will take another look, but this looks like it needs a bit of tweaking.

Comment (MuckT) [reply src/tokens/selectors.ts:69]: Added some clarifying comments in https://github.com/valora-inc/wallet/pull/2645/commits/86fd596d2b008ea288201660048c4385bab05966!

File: src/components/TokenBalance.tsx [hunk 4]
```diff
@@ -21,27 +21,36 @@ import Colors from 'src/styles/colors'
...
 
-  if (tokensWithUsdValue.length === 0) {
-    // Don't show zero if we haven't fetched the tokens yet.
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    dispatch(fetchTokenBalances())
+  }, [])
+
+  if (tokenFetchError || tokenFetchLoading || tokensAreStale) {
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/components/TokenBalance.tsx:48]: One issue that comes to mind with removing the tokenFetchLoading is that after restoring an account we would temporarily display - $0.00 on initial load which is reason enough not to remove it. A follow up PR might be a better place for this fix. Maybe in [Homefeed improvement: make total balance in sync with transaction feed - RET 322](https://linear.app/valora/issue/RET-322/homefeed-improvement-make-total-balance-in-sync-with-transaction-feed)?

File: src/components/TokenBalance.tsx [hunk 5]
```diff
@@ -21,27 +21,36 @@ import Colors from 'src/styles/colors'
...
   const tokenFetchLoading = useSelector(tokenFetchLoadingSelector)
   const tokenFetchError = useSelector(tokenFetchErrorSelector)
+  const tokensAreStale = useSelector(stalePriceSelector)
 
-  if (tokensWithUsdValue.length === 0) {
-    // Don't show zero if we haven't fetched the tokens yet.
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    dispatch(fetchTokenBalances())
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/components/TokenBalance.tsx]: Removed in [b17d8bf](https://github.com/valora-inc/wallet/pull/2645/commits/b17d8bfa78573088d35a294598e43c0066134808) and tests updated in https://github.com/valora-inc/wallet/pull/2645/commits/96cf9bd663690423285257c5a3797595fe1129ee!

File: src/tokens/selectors.ts [hunk 6]
```diff
@@ -50,6 +50,25 @@ export const tokensWithUsdValueSelector = createSelector(tokensListSelector, (to
...
+export const stalePriceSelector = createSelector(tokensListSelector, (tokens) => {
+  // If no tokens then prices cannot be stale
+  if (tokens.length === 0) return false
+  // Put tokens with usdPrice into an array
+  const tokensWithUsdValue = tokens.filter((tokenInfo) => tokenInfo.usdPrice !== null)
+  // If tokens with usd value exist, check the time price was fetched and if ANY are stale - return true
+  // Else tokens usd values are not present so we know prices are stale - return true
+  if (tokensWithUsdValue.length > 0) {
+    return (
+      tokensWithUsdValue.filter(
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/tokens/selectors.ts]: Fixed in https://github.com/valora-inc/wallet/pull/2645/commits/967995587780823591be56f9474621f91882fd02; This does look much cleaner! 🧼

**Review by kathaypacific — APPROVED** [review]
yay thank you!! 🚢

**Comment by ValoraQA** [comment]
Hey @MuckT, We have verified above task on latest Android Internal Release Build V1.37.0 & updated the status in the test rails.

- When user open the application Total Balance is displayed as "$-" while balance is loading.

**Test rail links:**
**Android 10:** https://valoraapp.testrail.io/index.php?/tests/view/54616

**Devices:** Google Pixel XL (10.0)
 Thanks..!!

**Comment by ValoraQA** [comment]
Hey @MuckT, We have verified above task on Latest iOS Test flight Release build V1.37.0 & updated the status in the test rails.

**Observation**:
- When user open the application Total Balance is displayed as "$-" while balance is loading.

**Test rail link:**
**iOS 15:** https://valoraapp.testrail.io/index.php?/tests/view/54661

**Device**: iPhone 13 mini(15.1.1)

Thanks..!!

**Pull request merged**